js/canvas.ts

```diff
@@ -1010,47 +988,69 @@ export class VideoModes2D implements VideoModes {
     }
 
     mixed(on: boolean) {
-        const old = mixedMode;
-        mixedMode = on;
+        const old = this.mixedMode;
+        this.mixedMode = on;
         if (old != on) {
             this._refresh();
         }
     }
 
     page(pageNo: pageNo) {
-        const old = pageMode;
-        pageMode = pageNo;
+        const old = this.pageMode;
+        this.pageMode = pageNo;
         if (old != pageNo) {
             this._refresh();
         }
     }
 
     isText() {
-        return textMode;
+        return this.textMode;
     }
 
     isMixed() {
-        return mixedMode;
+        return this.mixedMode;
     }
 
     isPage2() {
-        return pageMode == 2;
+        return this.pageMode == 2;
     }
 
     isHires() {
-        return hiresMode;
+        return this.hiresMode;
     }
 
     isDoubleHires() {
-        return doubleHiresMode;
+        return this.doubleHiresMode;
     }
 
     is80Col() {
-        return _80colMode;
+        return this._80colMode;
     }
 
     isAltChar() {
-        return altCharMode;
+        return this.altCharMode;
+    }
+
+    buildScreen(mainData: ImageData, mixData?: ImageData | null) {
+        if (!tmpContext) {
+            throw new Error('No 2d context');
+        }
+
+        const { width, height } = { width: 560, height: 192 };
+        const { x, y } = this._80colMode ? { x: 0, y: 0 } : { x: 0, y: 0 };
+
+        tmpCanvas.width = width;
```

Comment:
The `tmpCanvas` is resized, cleared, and the fill style is reset every time. Should each instance have its own canvas and context? Does the clearing _really_ need to happen? I _think_ that if the mainData and mixData were build with an opaque black background, it wouldn't be necessary. If I [read it right](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createImageData), createImageData() always clears with solid black transparent, so any black drawn would have to be explicitly set to opaque. But that's exactly what the constructors do with the `imageData`. Maybe I'm missing something, though, and you have to do all this each time.

Reply:
I think you are right, I can clean this up.